### ISSUE TITLE
Rule Proposal: require-unicode-regexp

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#9961]
**Please describe what the rule should do:**

This rule enforces people to use RegExp `u` flag.

Because of [Annex B.1.4 Regular Expressions Patterns](https://www.ecma-international.org/ecma-262/8.0/#sec-regular-expressions-patterns), regular expressions have different between `u` flag or not. For example:

```js
/a{/.test("a{12}")     //→ true
/a{/u.test("a{12}")    //→ Incomplete quantifier
/a{1/.test("a{12}")    //→ true
/a{1/u.test("a{12}")   //→ Incomplete quantifier
/a{12/.test("a{12}")   //→ true
/a{12/u.test("a{12}")  //→ Incomplete quantifier
/a{12}/.test("a{12}")  //→ false
/a{12}/u.test("a{12}") //→ false
/\x/.test("x")         //→ true
/\x/u.test("x")        //→ Invalid escape
```

This is similar to `"use strict"`.
The `u` flag makes clearer syntax and helps to catch mistakes early.

**What category of rule is this? (place an "X" next to just one item)**

[X] Enforces code style
[ ] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
//✘ BAD
/foo/

//✓ GOOD
/foo/u
```

**Why should this rule be included in ESLint (instead of a plugin)?**

We have `strict` rule for `"use strict"`.
This rule is similar to the `strict` rule, disabling legacy syntax to catch mistakes early.

**Comment by ljharb** [comment]
I like the rule, but I'm not sure the name is appropriate - "strict" is pretty broad.

What about `require-unicode-regexp`?

**Comment by mysticatea** [comment]
Sounds good to me.

**Comment by not-an-aardvark** [comment]
This issue is now accepted.
